2019-11-storybook-for-web-components-on-steroids/README.md

```diff
@@ -0,0 +1,439 @@
+function _defineProperties(target, props) {
+// ... more helper functions
+
+var MyClass =
+/*#__PURE__*/
+function () {
+  function MyClass() {
+    _classCallCheck(this, MyClass);
+
+    _privateField.set(this, {
+      writable: true,
+      value: 'My Class with a private field'
+    });
+  }
+
+  _createClass(MyClass, [{
+    key: "publicMethod",
+    value: function publicMethod() {
+      document.body.innerHTML = _classPrivateFieldGet(this, _privateField);
+      debugger;
+    }
+  }]);
+
+  return MyClass;
+}();
+
+var _privateField = new WeakMap();
+```
+
+Wow that is quite some code 🙈 and it does not really look like the code written 😱
+
+what happens? in a typical webpack & babel setup your code gets compiled down to es5 in order to be able to run the code also on older browser like IE11.
+
+However you may ask how often do I actually run my app in an older browser?
+
+A typical developer should probably develop ~90% on a modern browser and ~10% on older browsers to make sure everything still works in order.
+At least we hope you have such a nice workflow 🤗
+
+So the question is why compile, ship, debug and work with this "strange" code 100% of the time if it's only needed for 10%?
+Could we do better?
+
+Lets see how `es-dev-server` handles it by opening the same file on chrome.
+
+```js
+export class MyClass {
+  #privateField = 'My Class with a private field';
+
+  publicMethod() {
+    document.body.innerHTML = this.#privateField;
+    debugger;
+  }
+}
+```
+
+It looks exactly as the original code - because it is. The code as is was fully capable to run in chrome without any adjustments.
+And that's what is happening it ships the source as is.
+
+However we are using private class fields an unsupported feature by for example Firefox.
+What happens if we open it there?
+
+it fails 😭
+
+> SyntaxError: private fields are not currently supported
+
+ok, it's our fault as we are using a stage 3 feature and are not doing any compilation now.
+
+so let's do `es-dev-server --babel` which now uses the same `.babelrc` as webpack.
+
+```js
+function _classPrivateFieldGet(receiver, privateMap) {
+  var descriptor = privateMap.get(receiver);
+  if (!descriptor) {
+    throw new TypeError('attempted to get private field on non-instance');
+  }
+  if (descriptor.get) {
+    return descriptor.get.call(receiver);
+  }
+  return descriptor.value;
+}
+
+export class MyClass {
+  constructor() {
+    _privateField.set(this, {
+      writable: true,
+      value: 'My Class with a private field',
+    });
+  }
+
+  publicMethod() {
+    document.body.innerHTML = _classPrivateFieldGet(this, _privateField);
+    debugger;
+  }
+}
+
+var _privateField = new WeakMap();
+```
+
+So now it work 💪
+It only compiles the private fields and not everything 👍
+
+However if you now go back to chrome you will see that it is now compiled there as well...
```

Comment:
@LarsDenBakker that was a little surprising to me... is it possible to do more magic to not compile on chrome if not needed and compile on Firefox if needed? 😬

Reply:
This is with a custom babel config? Yeah we dont know whats inside it and what rules to apply. The auto compile detection only works for stage 4 things